src/main/java/org/jabref/logic/importer/util/StaxParser.java

```diff
@@ -0,0 +1,128 @@
+
+                // append the start tag
+                content.append(getXMLStartTag(reader, false));
+            } else if (event == XMLStreamConstants.END_ELEMENT) {
+                String tagName = reader.getLocalName();
+
+                // append the end tag
+                content.append(getXMLEndTag(reader));
+
+                if (tagName.equals(parentTag)) {
+                    depth--;
+
+                    if (depth == 0) {
+                        // reached the closing tag of the first parent tag
+                        break;
+                    }
+                }
+            } else if (event == XMLStreamConstants.CHARACTERS) {
+                content.append(getXMLText(reader));
+            } else if (event == XMLStreamConstants.CDATA) {
+                content.append(getXMLCData(reader));
+            } else if (event == XMLStreamConstants.COMMENT) {
+                content.append(getXMLComment(reader));
+            } else if (event == XMLStreamConstants.PROCESSING_INSTRUCTION) {
+                content.append(getXMLProcessingInstruction(reader));
+            } else if (event == XMLStreamConstants.SPACE || event == XMLStreamConstants.ENTITY_REFERENCE) {
+                content.append(getXMLText(reader));
+            }
+        }
+
+        return content.toString().trim();
+    }
+
+    private static String getXMLStartTag(XMLStreamReader reader, boolean addNamespaceURI) {
+        StringBuilder startTag = new StringBuilder();
+
+        String prefix = reader.getPrefix();
+
+        startTag.append("<")
+                .append(prefix != null && !prefix.isBlank() ? prefix + ":" : "")
+                .append(reader.getName().getLocalPart());
+
+        String namespaceURI = reader.getNamespaceURI();
+        if (addNamespaceURI && namespaceURI != null) {
+            startTag.append(" xmlns")
+                    .append(prefix != null && !prefix.isBlank() ? ":" + prefix : "")
+                    .append("=\"")
+                    .append(namespaceURI)
+                    .append("\"");
+        }
+
+        for (int i = 0; i < reader.getAttributeCount(); i++) {
+            startTag.append(" ").append(reader.getAttributeLocalName(i)).append("=\"").append(reader.getAttributeValue(i)).append("\"");
+        }
+
+        if (reader.isEndElement()) {
+            startTag.append("/");
+        }
+
+        startTag.append(">");
+        return startTag.toString();
+    }
+
+    private static String getXMLEndTag(XMLStreamReader reader) {
+        StringBuilder endTag = new StringBuilder();
+        String prefix = reader.getPrefix();
+
+        endTag.append("</")
+              .append(prefix != null && !prefix.isBlank() ? prefix + ":" : "")
+              .append(reader.getName().getLocalPart())
+              .append(">");
+
+        return endTag.toString();
+    }
+
+    private static String getXMLCData(XMLStreamReader reader) {
+        return "<![CDATA[" + reader.getText() + "]]>";
```

Comment:
Honestly, I don't understand this class. What is the purpose of building xml tags manually again?

Reply:
This class was added since we are using a StaX parser in the `MedlineImporter` class, which does not load the full XML data into memory. Instead we have a stream and can progress only in a forward manner traversing through all the tag elements. I was not able to find any inbuilt library method on the StaX parser itself to easily extract the content between two main parent tags (`<math>` in this case), so this custom logic was required. Following that, the extracted XML string is used for carrying out the transformation. I hope that clarifies things!

Reply:
Ah yeah I see, thanks for the explanation. Searched a bit around but seems like the Stax Parser is only for single documents. So this is fine for me then!